src/x25519.rs

```diff
@@ -12,13 +12,93 @@
 //! This implements x25519 key exchange as specified by Mike Hamburg
 //! and Adam Langley in [RFC7748](https://tools.ietf.org/html/rfc7748).
 
+use core::ops::Mul;
+
+use clear_on_drop::clear::Clear;
+
 use curve25519_dalek::constants::ED25519_BASEPOINT_TABLE;
 use curve25519_dalek::montgomery::MontgomeryPoint;
 use curve25519_dalek::scalar::Scalar;
 
 use rand_core::RngCore;
 use rand_core::CryptoRng;
 
+/// A DH ephemeral public key.
+#[repr(C)]
+pub struct EphemeralPublic(pub (crate) MontgomeryPoint);
+
+/// A DH ephemeral secret key.
+#[repr(C)]
+#[derive(Default)] // we derive Default in order to use the clear() method in Drop
+pub struct EphemeralSecret(pub (crate) Scalar);
+
+/// Overwrite ephemeral secret key material with null bytes when it goes out of scope.
+impl Drop for EphemeralSecret {
+    fn drop(&mut self) {
+        self.0.clear();
+    }
+}
+
+/// Multiply this `EphemeralPublic` key by a `EphemeralSecret` key.
+impl<'a, 'b> Mul<&'b EphemeralSecret> for &'a EphemeralPublic {
+    type Output = EphemeralPublic;
+
+    fn mul(self, secret: &'b EphemeralSecret) -> EphemeralPublic {
+        EphemeralPublic(self.0 * secret.0)
+    }
+}
+
+impl EphemeralSecret {
+    /// Utility function to make it easier to call `x25519()` with
+    /// an ephemeral secret key and montegomery point as input and
+    /// a shared secret as the output.
+    pub fn diffie_hellman(&self, their_public: &EphemeralPublic) -> SharedSecret {
+        SharedSecret(x25519(&self.0, &MontgomeryPoint(*their_public.0.as_bytes())))
+    }
+
+    /// Generate an x25519 `EphemeralSecret` key.
+    pub fn new<T>(csprng: &mut T) -> Self
+        where T: RngCore + CryptoRng
+    {
+        let mut bytes = [0u8; 32];
+
+        csprng.fill_bytes(&mut bytes);
+
+        EphemeralSecret(decode_scalar(&bytes))
+    }
+
+}
+
+impl<'a> From<&'a EphemeralSecret> for EphemeralPublic {
+    /// Given an x25519 `EphemeralSecret` key, compute its corresponding
+    /// `EphemeralPublic` key.
+    fn from(secret: &'a EphemeralSecret) -> EphemeralPublic {
+        EphemeralPublic((&ED25519_BASEPOINT_TABLE * &secret.0).to_montgomery())
+    }
+
+}
+
+/// A DH SharedSecret
+#[repr(C)]
+#[derive(Default)] // we derive Default in order to use the clear() method in Drop
```

Comment:
Is it really necessary to derive `Default` (instead of just calling `self.0.clear()` in the `Drop` impl)?

It seems like it would be better not to implement `Default`, so that the only way to create a `SharedSecret` is as the output of a DH function.